cpp/include/cudf/scalar/scalar.hpp

```diff
@@ -0,0 +1,267 @@
+   static_assert(is_fixed_width<T>(), "Unexpected non-fixed-width type.");
+ public:
+  using value_type = T;
+
+  ~fixed_width_scalar() = default;
+  fixed_width_scalar(fixed_width_scalar&& other) = default;
+  fixed_width_scalar(fixed_width_scalar const& other) = default;
+
+  /**
+   * @brief Set the value of the scalar
+   * 
+   * @param value New value of scalar
+   * @param stream The CUDA stream to do the operation in
+   */
+  void set_value(T value, cudaStream_t stream = 0) { 
+    _data.set_value(value, stream);
+    this->set_valid(true, stream);
+  }
+
+  /**
+   * @brief Get the value of the scalar
+   * 
+   * @param stream The CUDA stream to do the operation in
+   */
+  T value(cudaStream_t stream = 0) { return _data.value(stream); }
+
+  /**
+   * @brief Returns a raw pointer to the value in device memory
+   */
+  T* data() { return _data.data(); }
+
+ protected:
+  rmm::device_scalar<T> _data{};  ///< device memory containing the value
+
+  fixed_width_scalar() = default;
+
+  /**
+   * @brief Construct a new fixed width scalar object
+   * 
+   * @param value The initial value of the scalar
+   * @param is_valid Whether the value held by the scalar is valid
+   * @param stream The CUDA stream to do the allocation in
+   * @param mr The memory resource to use for allocation
+   */
+  fixed_width_scalar(T value, bool is_valid = true, cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
+   : scalar(data_type(experimental::type_to_id<T>()), is_valid, stream, mr)
+   , _data(value, stream, mr)
+  {}
+};
+
+} // namespace detail
+
+
+/**
+ * @brief An owning class to represent a numerical value in device memory
+ * 
+ * @tparam T the data type of the numerical value
+ */
+template <typename T>
+class numeric_scalar : public detail::fixed_width_scalar<T> {
+   static_assert(is_numeric<T>(), "Unexpected non-numeric type.");
+ public:
+  numeric_scalar() = default;
+  ~numeric_scalar() = default;
+  numeric_scalar(numeric_scalar&& other) = default;
+  numeric_scalar(numeric_scalar const& other) = default;
+
+  /**
+   * @brief Construct a new numeric scalar object
+   * 
+   * @param value The initial value of the scalar
+   * @param is_valid Whether the value held by the scalar is valid
+   * @param stream The CUDA stream to do the allocation in
+   * @param mr The memory resource to use for allocation
+   */
+  numeric_scalar(T value, bool is_valid = true, cudaStream_t stream = 0,
+      rmm::mr::device_memory_resource *mr = rmm::mr::get_default_resource())
+   : detail::fixed_width_scalar<T>(value, is_valid, stream, mr)
+  {}
+};
+
+/**
+ * @brief An owning class to represent a string in device memory
+ */
+class string_scalar : public scalar {
+ public:
+  using value_type = cudf::string_view;
+
+  string_scalar() = default;
+  ~string_scalar() = default;
+  string_scalar(string_scalar&& other) = default;
+  string_scalar(string_scalar const& other) = default;
```

Comment:
It is better to explicit about the fact that we are deleting the move/copy assignment operators here (as well as in all the other derived classes) https://en.cppreference.com/w/cpp/language/rule_of_three

Reply:
Won't the base class' deleted methods prevent the generation of these in the derived classes?

Reply:
It isn't the base class' deleted methods that are preventing the generation, see here (the assignment operators are still not compiler generated): https://godbolt.org/z/B6kEb1

It is the fact that you are **explicitly** declaring your copy/move constructors in the base & derived class as default. If you change the explicit to implicit, then they are generated, see here: https://godbolt.org/z/OSSKaW

Keeping track of which SMF (special member functions) a compiler generates is extremely confusing (see the chart by Howard Hinnant below). My rule of thumb is: if you are doing anything more complicated that default constructor / destructor or custom constructor / default descuctor, **be explicit about every SMF** - otherwise the reader can't be sure of what the author intended.

![image](https://user-images.githubusercontent.com/36027403/68057391-1b1e2880-fccc-11e9-9a2c-7e8e1eb9b337.png)
https://foonathan.net/2019/02/special-member-functions/